singleton-service/src/service/proxy.rs

```diff
@@ -0,0 +1,155 @@
+pub use crate::configuration::service::ServiceConfig;
+pub use crate::service::deltas::DeltaStore;
+use log::{debug, info, warn};
+use proxy_wasm::{
+    hostcalls::{dequeue_shared_queue, register_shared_queue},
+    traits::{Context, RootContext},
+    types::LogLevel,
+};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use threescale::{
+    proxy::cache::{get_application_from_cache, set_application_to_cache},
+    structs::ThreescaleData,
+    utils::update_metrics,
+};
+
+// QUEUE_NAME should be the same as the one in cache filter.
+const QUEUE_NAME: &str = "message_queue";
+
+#[no_mangle]
+pub fn _start() {
+    proxy_wasm::set_log_level(LogLevel::Trace);
+    proxy_wasm::set_root_context(|context_id| -> Box<dyn RootContext> {
+        Box::new(SingletonService {
+            context_id,
+            config: ServiceConfig::default(),
+            queue_id: None,
+            delta_store: DeltaStore {
+                last_update: None,
+                request_count: 0,
+                deltas: RefCell::new(HashMap::new()),
+            },
+        })
+    });
+}
+
+struct SingletonService {
+    context_id: u32,
+    config: ServiceConfig,
+    queue_id: Option<u32>,
+    delta_store: DeltaStore,
+}
+
+impl RootContext for SingletonService {
+    /// Message queue will get registered when on_vm_start callback gets called.
+    fn on_vm_start(&mut self, _vm_configuration_size: usize) -> bool {
+        if let Ok(q_id) = register_shared_queue(QUEUE_NAME) {
+            self.queue_id = Some(q_id);
+        }
+        // TODO : handle MQ failure, change info to trace after dev
```

Comment:
Failing to register proxy will cause the proxy to crash so you don't have to worry about this.

Reply:
The proxy should never crash - the host implementation might panic, which is a well known bad practice, and hopefully we'll avoid that in the future and provide a status from the SDK that can be helpful to log and diagnose the issue. IOW, just don't depend on code that you don't control (the SDK) but instead depend on the full signature (ie. it can be an `Err(...)`) - this way we can have the SDK fixed without having to hunt for implicit dependencies in our code.

Reply:
I'll write these design problems from the SDK side inside the design doc for later use.